### PR TITLE
udfore: basic parser (WIP)

### PR DESCRIPTION
**Pull request by Abb1x**
I started working on a basic parser for udfore, it supports:
- variable decl
- function decl
- type decl with tuples
- single constant
- references to variables
- function calls
- outputting to JSON

The code is most likely not good but it (kinda) works
still WIP

### CODE REVIEW

File: sources/libs/ud/parse/type.c [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+{
+    UdDecl ret = {};
+
+    ret.type = UD_DECL_TYPE;
+
+    ud_parse_whitespace(lex);
+
+    if (ud_expect(lex, UDLEX_IDENT))
+    {
+        ret.name = lex_peek(lex, -1).str;
```
Comment (Supercip971) [inline sources/libs/ud/parse/type.c:73]: maybe ?
```c
if(ud_curr_type(lex) == UDLEX_IDENT)
{
    ret.name = lex_curr(lex).str;
    lex_next();
}
```

File: sources/libs/ud/parse/type.c [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    ud_parse_whitespace(lex);
+
+    ud_expect(lex, UDLEX_EQUAL);
+
```
Comment (Supercip971) [inline sources/libs/ud/parse/type.c:79]: maybe do like cc, add a ud_expect_spaced() (I don't remember the name) that is just;
```c
void ud_expect_spaced(Lex* lex, ClexType type)
{
     ud_parse_whitespace(lex);
     ud_expect(lex, UDLEX_EQUAL);
     ud_parse_whitespace(lex);
}
```

File: sources/libs/ud/parse/parse.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    switch (lexeme.type)
+    {
+    case UDLEX_INTEGER:
+        ret.type = UD_VAL_INT;
+        ret.int_ = str_to_number(lexeme.str);
+        break;
+    case UDLEX_STRING:
+        ret.type = UD_VAL_STR;
+        ret.string = lexeme.str;
```
Comment (Supercip971) [inline sources/libs/ud/parse/parse.c:22]: no break + no default ?

File: sources/libs/ud/parse/misc.c [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+bool ud_expect(Lex *lex, LexemeType type)
+{
+    if (lex_expect(lex, type))
+    {
+        return true;
+    }
+
+    print(embed_log_writer(), "Error while parsing: expected '{}' at {}:{}\n", udlex_to_str(type), lex_curr(lex).line, lex_curr(lex).col);
```
Comment (Supercip971) [inline sources/libs/ud/parse/misc.c:58]: ? log$() ?

File: sources/libs/ud/parse/type.c (hunk 2 above)
Comment (Supercip971) [reply sources/libs/ud/parse/type.c:79]: oh wait no you have ud_parse_expect_separator

File: sources/libs/ud/parse/parse.c [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+        if (out.expr.decl.type == UD_DECL_NONE)
+        {
+            out = ud_parse_expr(lex, alloc);
+
+            lex_next(lex);
+
+            ud_parse_whitespace(lex);
+
+            ud_expect(lex, UDLEX_SEMICOLON);
```
Comment (Supercip971) [inline sources/libs/ud/parse/parse.c:116]: ud_parse_expect_separator

File: sources/libs/ud/parse/parse.c [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+            ud_parse_whitespace(lex);
+        }
+
+        ret = out;
```
Comment (Supercip971) [inline sources/libs/ud/parse/parse.c:121]: return out ?

File: sources/libs/ud/parse/parse.c [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    if (lex_curr_type(lex) == LEXEME_EOF)
+    {
+        return ret;
+    }
+
+    if (lex_curr(lex).type != UDLEX_WHITESPACE && lex_curr(lex).type != UDLEX_COMMENT && lex_curr(lex).type != LEXEME_EOF)
+    {
+        UdAstNode out = ud_parse_decl(lex, alloc);
+
+        if (out.expr.decl.type == UD_DECL_NONE)
```
Comment (Supercip971) [inline sources/libs/ud/parse/parse.c:108]: at l.104 you shouldd sheck if it is a decl

File: sources/libs/ud/parse/parse.c [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    ud_parse_whitespace(lex);
+
+    UdExpr right = ud_parse_expr(lex, alloc).expr;
+
+    log$("{}", right.const_.int_);
+
+    ret.left = alloc_malloc(alloc, sizeof(UdExpr));
+    ret.right = alloc_malloc(alloc, sizeof(UdExpr));
+
+    mem_cpy(ret.left, &left, sizeof(UdExpr));
```
Comment (Supercip971) [inline sources/libs/ud/parse/parse.c:57]: *ret.left = left;

File: sources/libs/ud/parse/lexer.c [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+        while (!scan_skip(scan, '"') && !scan_ended(scan))
+        {
+            scan_next(scan);
+        }
+
+        return (UDLEX_STRING);
+    }
+
+    else if (scan_skip_word(scan, str$("->")))
```
Comment (Supercip971) [inline sources/libs/ud/parse/lexer.c:65]: should be in keyword.type

File: sources/libs/ud/parse/lexer.c [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    {UDLEX_COMMA, ","},
+    {UDLEX_LCBRACE, "{"},
+    {UDLEX_RCBRACE, "}"},
+    {UDLEX_DOUBLE_EQUAL, "=="},
+    {UDLEX_EQUAL, "="},
+    {UDLEX_SEMICOLON, ";"}};
+
+static int isidentchar(char v)
+{
+    return isalpha(v) || v == '_' || v == '$' || v == '@' || isdigit(v);
```
Comment (Supercip971) [inline sources/libs/ud/parse/lexer.c:30]: '$' should be a separated token

File: sources/libs/ud/parse/lexer.c [hunk 11]
```diff
@@ -0,0 +1,116 @@
+#include <ud/parse/lexer.h>
+
+struct
+{
+    LexemeType type;
+    char const *literal;
+} keywords[] = {
+    {UDLEX_LET, "let "},
```
Comment (Supercip971) [inline sources/libs/ud/parse/lexer.c:8]: no space in keyword :^(

File: sources/libs/ud/parse/func.c [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+                ud_parse_whitespace(lex);
+            }
+        }
+
+        else
+        {
+            ud_parse_whitespace(lex);
+
+            lex_next(lex);
```
Comment (Supercip971) [inline sources/libs/ud/parse/func.c:133]: ud_parse_next_separator

File: sources/libs/ud/parse/decl.c [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+#include <ud/parse/lexer.h>
+
+UdAstNode ud_parse_decl(Lex *lex, Alloc *alloc)
+{
+    UdAstNode ret = {};
+
+    ret.expr.type = UD_EXPR_DECL;
+    ret.expr.decl.type = UD_DECL_NONE;
+
+    if (lex_expect(lex, UDLEX_LET))
```
Comment (Supercip971) [inline sources/libs/ud/parse/decl.c:11]: switch ? maybe it's too small for a switch

File: sources/libs/ud/gen/json.c [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+    case UD_EXPR_REFERENCE:
+    {
+        ud_emit_reference(expr.reference, json, alloc);
+        break;
+    }
+
+    case UD_EXPR_FUNC_CALL:
+    {
+        ud_emit_func_call(expr.func_call, json, alloc);
```
Comment (Supercip971) [inline sources/libs/ud/gen/json.c:204]: break;

File: sources/libs/brutal/parse/lex.h [hunk 15]
```diff
@@ -19,6 +22,8 @@ typedef struct
 
     bool has_error;
     LexError error;
+
+    int line, col;
```
Comment (Supercip971) [inline sources/libs/brutal/parse/lex.h:26]: int line;
int col;

Comment (Supercip971) [reply sources/libs/brutal/parse/lex.h:26]: maybe ?

**Review by Supercip971 — CHANGES_REQUESTED** [review]
WIP review, but looks good imo

**Comment by Supercip971** [comment]
maybe squash some commits ?

File: sources/libs/ud/parse/type.c (hunk 1 above)
Comment (Abb1x) [reply sources/libs/ud/parse/type.c:73]: yes but I won't get error handling. I think I'll put a Maybe as the return type

File: sources/libs/ud/repl/repl.c [hunk 16]
```diff
@@ -0,0 +1,31 @@
+#include <ud/repl.h>
+
+#ifdef USE_READLINE
+// clang-format off
```
[the remark this answers is not in the document]
Comment (Abb1x) [reply sources/libs/ud/repl/repl.c:4]: yes

**Pull request closed without merging**